crates/bevy_app/src/preferences.rs

```diff
@@ -0,0 +1,173 @@
+use bevy_ecs::system::Resource;
+use bevy_reflect::{Map, Reflect, TypePath};
+
+use crate::Plugin;
+
+/// Adds application [`Preferences`] functionality.
+pub struct PreferencesPlugin;
+
+impl Plugin for PreferencesPlugin {
+    fn build(&self, app: &mut crate::App) {
+        app.init_resource::<Preferences>();
+    }
+}
+
+/// A map that stores all application preferences.
+///
+/// Preferences are strongly typed, and defined independently by any `Plugin` that needs persistent
+/// settings. Choice of serialization format and behavior is up to the application developer. The
+/// preferences resource simply provides a common API surface to consolidate preferences for all
+/// plugins in one location.
+///
+/// ### Usage
+///
+/// Preferences only require that the type implements [`Reflect`].
+///
+/// ```
+/// # use bevy_reflect::Reflect;
+/// #[derive(Reflect)]
+/// struct MyPluginPreferences {
+///     do_things: bool,
+///     fizz_buzz_count: usize
+/// }
+/// ```
+/// You can [`Self::get`] or [`Self::set`] preferences by accessing this type as a [`Resource`]
+/// ```
+/// # use bevy_ecs::prelude::*;
+/// # use bevy_app::Preferences;
+/// # use bevy_reflect::Reflect;
+/// #
+/// # #[derive(Reflect)]
+/// # struct MyPluginPreferences {
+/// #     do_things: bool,
+/// #     fizz_buzz_count: usize
+/// # }
+/// #
+/// fn update(mut prefs: ResMut<Preferences>) {
+///     let settings = MyPluginPreferences {
+///         do_things: false,
+///         fizz_buzz_count: 9000,
+///     };
+///     prefs.set(settings);
+///
+///     // Accessing preferences only requires the type:
+///     let mut new_settings = prefs.get::<MyPluginPreferences>().unwrap();
+///
+///     // If you are updating an existing struct, all type information can be inferred:
+///     new_settings = prefs.get().unwrap();
+/// }
+/// ```
+#[derive(Resource, Default, Debug)]
```

Comment:
I believe `Preferences` needs some kind of de/serialisation support to really be useful. Ideally, a generic `FilePreferencesStorePlugin` could store/load the entire preferences resource to disk, as an example.

Reply:
That's why this is based on reflect. You can serialize the map however you'd like, in any format you'd like. I can't implement reflect on `Preferences` as-is, but it possible with some elbow grease for the internal `map`.

Reply:
I proved this out in a test:

```rs
    #[test]
    fn serialization() {
        let mut preferences = Preferences::default();
        preferences.set(Foo(42));
        preferences.set(Bar("Bevy".into()));

        for (_k, value) in preferences.map.iter() {
            let registry = get_registry();
            let serializer = bevy_reflect::serde::ReflectSerializer::new(value, &registry);
            let mut buf = Vec::new();
            let format = serde_json::ser::PrettyFormatter::with_indent(b"    ");
            let mut ser = serde_json::Serializer::with_formatter(&mut buf, format);
            use serde::Serialize;
            serializer.serialize(&mut ser).unwrap();

            let output = std::str::from_utf8(&buf).unwrap();
            println!("{:#}", output);
        }
    }
```

which prints

```
{
    "bevy_app::preferences::tests::Foo": [
        42
    ]
}
{
    "bevy_app::preferences::tests::Bar": [
        "Bevy"
    ]
}
```

Reply:
That makes sense, thanks for the clarification! For now, could you mark the inner `map` as `pub`? That way if we don't get a follow-up PR we can still get utility out of this?

Reply:
I've added this test and included fully roundtripping a file. The test uses JSON, but you could use whatever format you'd like.

Reply:
I've provided `iter_reflect`, which gives all the data needed to serialize the data, and `set_dyn` to deserialize the data, without needing to make the inner `map` `pub`.

Reply:
Someone smarter than me can figure out how to implement reflect, because it will require an assembly step to convert between the list of trait objects in the file, to a map in memory.

Reply:
> I can't implement reflect on `Preferences` as-is, but it possible with some elbow grease for the internal `map`.

Have you tried this?

```suggestion
#[derive(Resource, Reflect, Default, Debug)]
#[reflect(from_reflect = false)]
```

`DynamicMap` doesn't implement `FromReflect`, so you'll have to opt out of `FromReflect` for `Preferences`, but it should still be possible.

Note that when deserializing `Preferences`, you'll get back a `DynamicStruct`. Since you can't use `FromReflect` to convert it back to `Preferences`, you might need to use `Default`/`ReflectDefault` and patch with `Reflect::apply`.

Reply:
This is pretty outside of my comfort zone with `Reflect`. I had a sense that I needed to do something along those lines, but I have no idea what they do or what tools to reach for. If you can find a way to do that, I would appreciate a PR against this one!